tests/fixture-valid.js

```diff
@@ -150,34 +157,153 @@ function checkPhysical(physical, modeDescription = '') {
+    return;
+  }
+
+  const matrixJson = matrix.jsonObject;
+  const hasPixelCount = 'pixelCount' in matrixJson;
+  const hasPixelKeys = 'pixelKeys' in matrixJson;
+
+  if (!hasPixelCount && !hasPixelKeys) {
+    result.errors.push('Matrix must either define \'pixelCount\' or \'pixelKeys\'.');
+    return;
+  }
+  if (hasPixelCount && hasPixelKeys) {
+    result.errors.push('Matrix can\'t define both \'pixelCount\' and \'pixelKeys\'.');
+    return;
+  }
+
+  if (matrix.definedAxes.length === 0) {
+    result.errors.push('Matrix may not consist of only a single pixel.');
+    return;
+  }
+
+  const variesInAxisLength = matrix.pixelKeyStructure.some(
+    rows => rows.length !== matrix.pixelCountY || rows.some(
+      columns => columns.length !== matrix.pixelCountX
+    )
+  );
+  if (variesInAxisLength) {
+    result.errors.push('Matrix must not vary in axis length.');
+  }
+
+  checkPixelGroups(matrix);
+}
+
+/**
+ * Check if the referenced pixelKeys from the pixelGroups exist and are not referenced more than once.
+ * @param {Matrix} matrix The Matrix instance
+ */
+function checkPixelGroups(matrix) {
+  for (const pixelGroupKey of matrix.pixelGroupKeys) {
+    const usedMatrixChannels = new Set();
+
+    if (matrix.pixelKeys.includes(pixelGroupKey)) {
+      result.errors.push(`pixelGroupKey '${pixelGroupKey}' is already used as pixelKey. Please choose a different name.`);
+    }
+
+    for (const pixelKey of matrix.pixelGroups[pixelGroupKey]) {
+      if (!matrix.pixelKeys.includes(pixelKey)) {
+        result.errors.push(`pixelGroup '${pixelGroupKey}' references unknown pixelKey '${pixelKey}'.`);
+      }
+      if (usedMatrixChannels.has(pixelKey)) {
+        result.errors.push(`pixelGroup '${pixelGroupKey}' can't reference pixelKey '${pixelKey}' more than once.`);
+      }
+      usedMatrixChannels.add(pixelKey);
+    }
+  }
+}
+
+/**
+ * Check if templateChannels are defined correctly. Does not check the channel data itself.
+ * @param {!object} fixtureJson The fixture's JSON data
+ */
+function checkTemplateChannels(fixtureJson) {
+  if (isNotEmpty(fixtureJson.templateChannels, 'templateChannels are empty. Add a channel or remove it.')) {
+    if (fixture.matrix === null) {
+      result.errors.push('templateChannels are defined but matrix data is missing.');
+      return;
+    }
+
+    for (const templateChannel of fixture.templateChannels) {
+      checkTemplateChannel(templateChannel);
+    }
+  }
+  else if (fixture.matrix !== null) {
+    result.errors.push('Matrix is defined but templateChannels are missing.');
+  }
+}
+
+/**
+ * Check if the templateChannel is defined correctly. Does not check the channel data itself.
+ * @param {!TemplateChannel} templateChannel The templateChannel to examine.
+ */
+function checkTemplateChannel(templateChannel) {
+  checkTemplateVariables(templateChannel.name, ['$pixelKey']);
+
+  for (const key of templateChannel.allTemplateKeys) {
+    checkTemplateVariables(key, ['$pixelKey']);
```

Comment:
Shouldn't we test the name here, too? (Instead of above)

Reply:
No. `allTemplateKeys` are the current key + all aliases/keys that reference other template channels. Aliases don't have special names and keys reference other template channels that are checked before/later as well.